cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -278,34 +278,39 @@ struct CastFunctor<O, I, typename std::enable_if<is_float_truncate<O, I>::value>
     const in_type* in_data = GetValues<in_type>(input, 1);
     auto out_data = GetMutableValues<out_type>(output, 1);
 
-    if (options.allow_float_truncate) {
-      // unsafe cast
+    constexpr int64_t kMax = 1LL << std::numeric_limits<in_type>::digits;
+    constexpr int64_t kMin = -kMax;
+
+    if (input.null_count != 0) {
+      internal::BitmapReader is_valid_reader(input.buffers[0]->data(), in_offset,
+                                             input.length);
       for (int64_t i = 0; i < input.length; ++i) {
-        *out_data++ = static_cast<out_type>(*in_data++);
+        auto out_value = static_cast<out_type>(*in_data);
+        if (!options.allow_float_truncate &&
+            ARROW_PREDICT_FALSE(static_cast<in_type>(out_value) != *in_data)) {
+          ctx->SetStatus(Status::Invalid("Floating point value truncated"));
+        }
+        if (!options.allow_float_overflow &&
+            ARROW_PREDICT_FALSE(out_value >= kMax || out_value <= kMin)) {
```

Comment:
Hm, I'm not sure this is totally legit since the conversion to `out_value` was lossy. Maybe better to compare `*in_data` with the maximum / minimum representable floating point integer?